### PR TITLE
[FEATURE] [MER-293] GE Cloud Mode for DataContext

### PR DESCRIPTION
**Pull request by roblim**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Updated `DataContext` to allow `ge_cloud_mode` flag
- `DataContext` instantiated with `ge_cloud_mode=True` will behave differently:
  - rather than relying on `great_expectations.yml` for a project's configuration, the context will rely on a default template defined at `great_expectations/data_context/templates.py:160`, which means the context's config will only exist in memory
  - this template contains various substitution variables using ${VAR} notation. The values are substituted on the fly from the following places, in order of precedence:
    - passed into the `DataContext` constructor at runtime
    - environment variables
    - a global configuration file located at `~/.great_expectations/great_expectations.conf`
    - a global configuration file located at `/etc/great_expectations.conf`

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [ ] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [ ] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [ ] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by roblim** [comment]
@anthonyburdi - just flagging you on this PR in case you wanted to preview

**Comment by roblim** [comment]
See additional PR notes here: https://superconductive.atlassian.net/l/c/M65cxC3h

File: great_expectations/data_context/data_context.py [hunk 1]
```diff
@@ -379,6 +397,14 @@ def __init__(self, project_config, context_root_dir=None, runtime_environment=No
...
         self._evaluation_parameter_dependencies = {}
 
+    @property
+    def ge_cloud_config(self):
+        return self._ge_cloud_config
+
+    @property
+    def ge_cloud_mode(self):
+        return self._ge_cloud_mode
+
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context.py:407]: <3

File: great_expectations/data_context/data_context.py [hunk 2]
```diff
@@ -379,6 +397,14 @@ def __init__(self, project_config, context_root_dir=None, runtime_environment=No
         self._evaluation_parameter_dependencies_compiled = False
         self._evaluation_parameter_dependencies = {}
 
+    @property
+    def ge_cloud_config(self):
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context.py:401]: ❤️

File: great_expectations/data_context/data_context.py [hunk 3]
```diff
@@ -3754,16 +3905,119 @@ def scaffold_notebooks(cls, base_dir):
...
+        ge_cloud_access_token: Optional[str] = None,
+    ):
+        ge_cloud_base_url = (
+            ge_cloud_base_url
+            or super()._get_global_config_value(
+                environment_variable="GE_CLOUD_BASE_URL",
+                conf_file_section="ge_cloud_config",
+                conf_file_option="base_url",
+            )
+            or "https://app.greatexpectations.io/"
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context.py:3922]: Just curious : How is `GE_CLOUD_BASE_URL` going to be configured for the Anthony and Sarah milestones? is `https://app.greatexpectations.io` part of either milestone?

File: great_expectations/data_context/data_context.py [hunk 4]
```diff
@@ -3754,16 +3905,119 @@ def scaffold_notebooks(cls, base_dir):
...
+                    f'context_root_dir was not provided - defaulting to current working directory "'
+                    f'{context_root_dir}".'
+                )
+        else:
+            # Determine the "context root directory" - this is the parent of "great_expectations" dir
+            context_root_dir = (
+                self.find_context_root_dir()
+                if context_root_dir is None
+                else context_root_dir
+            )
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context.py:4008]: ❤️

File: tests/data_context/fixtures/conf/great_expectations_cloud_config_complete_2.conf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+[ge_cloud_config]
+base_url: https://api.dev.greatexpectations.io/complete/version-2
+account_id: 31c84fc9-6659-4411-a911-4276bb464583
+access_token: 85327f94194e4b5b90db072029fcc474
+
+plugins_directory: /Users/foo/bar/my/plugins/directory/complete/version/2
+usage_statistics_url: https://dev.stats.greatexpectations.io/great_expectations/v1/usage_statistics/complete/version/2
```
Comment (Shinnnyshinshin) [inline tests/data_context/fixtures/conf/great_expectations_cloud_config_complete_2.conf:7]: Love these fixtures. Thank you

File: tests/data_context/conftest.py [hunk 6]
```diff
@@ -160,3 +162,163 @@ def basic_data_context_config():
...
+        new_callable=PropertyMock,
+    ) as mock:
+        (
+            mock_global_config_dot_dir,
+            mock_global_config_etc_dir,
+            mock_global_config_paths,
+        ) = mocked_global_config_dirs
+        mock.return_value = mock_global_config_paths
+
+        yield
```
Comment (Shinnnyshinshin) [inline tests/data_context/conftest.py:220]: Thank you for how these are organized

**Review by Shinnnyshinshin — COMMENTED** [review]
Thank you @roblim , I felt the PR was really well implemented, organized and tested... I also found the Notes section to be very helpful in getting my bearings.  I left a few notes and a small non-blocking question. 

I also had a question about which JIRA task was being used to track / document the changes in this PR. I tried looking for MER-293 and it took me to CLOUD-166 which didn't have that much detail. 

If it doesn't exist already, then can I ask that the Notes in this PR be added to a Confluence doc that is linked by the corresponding JIRA task? Or if I'm just not looking in the right place, please let me know. Thanks again :)

File: great_expectations/data_context/data_context.py (hunk 3 above)
Comment (roblim) [reply great_expectations/data_context/data_context.py:3922]: For the Anthony milestone, we'll just provide him with an appropriate `great_expectations.conf`, which would include the base url (in his case, https://app.dev.greatexpectations.io).

**Comment by roblim** [comment]
> Thank you @roblim , I felt the PR was really well implemented, organized and tested... I also found the Notes section to be very helpful in getting my bearings. I left a few notes and a small non-blocking question.
> 
> I also had a question about which JIRA task was being used to track / document the changes in this PR. I tried looking for MER-293 and it took me to CLOUD-166 which didn't have that much detail.
> 
> If it doesn't exist already, then can I ask that the Notes in this PR be added to a Confluence doc that is linked by the corresponding JIRA task? Or if I'm just not looking in the right place, please let me know. Thanks again :)

Sure thing. I've pulled the notes and moved them here (so there's single source of truth): https://superconductive.atlassian.net/l/c/M65cxC3h

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you @roblim. My questions have been addressed :) LGTM

**Pull request merged**